lib/elixir/lib/enum.ex

```diff
@@ -723,28 +723,30 @@ defmodule Enum do
   end
 
   def fetch(enumerable, index) when is_integer(index) and index < 0 do
-    case Enumerable.count(enumerable) do
+    module = Enumerable.impl_for!(enumerable)
+    case module.count(enumerable) do
       {:error, _module} ->
         enumerable |> reverse |> fetch_list((-index) - 1)
```

Comment:
Yes, we need to implement reverse by hand:

    enumerable |> module.reverse({:cont, []}, fn x, acc -> {:cont, [x | acc]} end) |> elem(1) |> fetch_list((-index) - 1)